add a callback for when a new extension is decoded

robguima opened the pull request:
👋Would typically use such a callback for fingerprinting the extensions.

I've added the CB from `decode_server_hello` and `client_handle_encrypted_extensions`, but should it perhaps also go into `decode_certificate_request`?

Does this makes sense in upstream? 
thanks!

kazuho commented on include/picotls.h:
```diff
@@ -496,6 +496,10 @@ PTLS_CALLBACK_TYPE(void, update_open_count, ssize_t delta);
+/**
+ * callback for every new extension detected during decoding of the client hello
+ */
+PTLS_CALLBACK_TYPE(void, on_new_client_extension, ptls_t *tls, uint16_t exttype);
```
How about changing the type to `PTLS_CALLBACK_TYPE(void, on_extension, ptls_t *tls, uint8_t message_type, uint16_t exttype, ptls_iovec_t extdata)`, and collecting all extensions regardless of the what the handshake message is?

The reasons are:
* it would be useful also for collecting extensions provided by servers when running as a client
* pass handshake message type so that we can collect extensions other than those found in ClientHello
* pass extension data as well, because passing a vector is basically free and it would give applications more freedom to do whatever they want

robguima replied:
Sounds great! Thanks for the suggestions! 
Should the name assume every extension cb would represent a new one? I take it extensions should not appear multiple times.

kazuho replied:
> Should the name assume every extension cb would represent a new one?

Yes, except the fact that picotls does not check that for extensions that it does not recognize. Also, I think you would receive some extensions (belonging to same handshake message) twice when HRR is involved.

robguima replied:
> picotls does not check that for extensions that it does not recognize.
oh well that would be desirable for fingerprinting.
OK. Don't think receiving the same one twice would be a problem: the CB can handle that (or ignore it).

robguima replied:
btw, do you mean `hstype`, as in handshake type defined in `init_extension_bitmap`, by message_type? should we keep it as `hstype` then?

kazuho replied:
> Don't think receiving the same one twice would be a problem: the CB can handle that (or ignore it).

Agreed.

> btw, do you mean `hstype`, as in handshake type defined in `init_extension_bitmap`, by message_type?

Correct. Naming it `hstype` works for me.

robguima commented:
6d46807 catches more extensions: not sure if too much or too little.
At times it is including the entire block, but changing that would mean further changes in the code.
the test in t/picotls.c is simplistic and `PTLS_HANDSHAKE_TYPE_ENCRYPTED_EXTENSIONS` is returning a empty length in the iovec, which seems to be OK based on https://github.com/h2o/picotls/blob/master/lib/picotls.c#L2313

kazuho commented on lib/picotls.c:
```diff
@@ -1006,7 +1006,7 @@ static int derive_resumption_secret(ptls_key_schedule_t *sched, uint8_t *secret,
 }
 
-static int decode_new_session_ticket(uint32_t *lifetime, uint32_t *age_add, ptls_iovec_t *nonce, ptls_iovec_t *ticket,
+static int decode_new_session_ticket(ptls_t *tls, uint32_t *lifetime, uint32_t *age_add, ptls_iovec_t *nonce, ptls_iovec_t *ticket,
```
Thank you for the changes.

I would appreciate it if you could use `ptls_context_t *` in the callbacks, because that helps using the function without creating `ptls_t *`. Other that I think the PR is ready to go.

robguima replied:
@kazuho thank you for the review!
Hmm the `on_extension` cb would need to rely on the ptls_t in order to tie to a particular connection. Perhaps we could skip ticket extension?
Or make these calls work with an without the ptls (e.g. NULL ptls), and always passing pctx in decode_stored_session_ticket?

kazuho replied:
> Hmm the `on_extension` cb would need to rely on the ptls_t in order to tie to a particular connection. Perhaps we could skip ticket extension?

Ah that's a good point. Let's keep the PR as-is. The static functions need not to be made complex; we can change them anytime. I was just wondering if the PR was making debatable changes that could be avoided. I now understand that I was wrong.

robguima replied:
cool! thanks!

kazuho commented on t/picotls.c:
```diff
@@ -510,6 +510,7 @@ static void test_fragmented_message(void)
 }
 
 static int was_esni;
+static int debug = 1;
```
I believe this is a leftover during your development?

Regarding the feature, I think I prefer having a test that always get executed, or having nothing at all.

kazuho commented on include/picotls.h:
```diff
@@ -496,6 +496,10 @@ PTLS_CALLBACK_TYPE(void, update_open_count, ssize_t delta);
+/**
+ * callback for every extension detected during decoding
+ */
+PTLS_CALLBACK_TYPE(void, on_extension, ptls_t *tls, uint8_t hstype, uint16_t exttype, ptls_iovec_t extdata);
```
Can we let the function return an `int` indicating the TLS alert codes? The caller would be expected to pass back the error value when anything other than 0 is returned.

robguima commented:
@kazuho thanks for the review and suggestions. Perhaps 8134f3e addresses your concerns: it lets the on_extension CB return a non-zero value/alert and returns it if non-zero.
Also cleaned up the leftover you spotted. The test is still fairly simple, and if can remove it if you prefer. Thought about doing something in `cli`, but not sure what would be of interest - perhaps the SHA of the extensions?

robguima commented:
I realize using this API/CB to build a sha/fingerprint would still demand storing something, the partial SHA, in the ptls structure. We don't have distincts initial, update, and final stages and the CB gets invoked as the extensions are decoded.

kazuho commented:
Thank you for the changes. Merged to master.
